### PR TITLE
Fix aircraft attack (strafe) runs with sliding behaviour

### PR DESCRIPTION
**Pull request by abcdefg30**
Having the `Default` or `Strafe` attack types together with `CanSlide: true` produced this
strafeSlide

This PR introduces `FlyDistance` which just flies the distance given in a straight line from the position the activity starts and then uses that new activity to fly the strafing distance instead of relying on edge cases inside the default `Fly`.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/Air/FlyDistance.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyDistance.cs]: What do you think about calling this `FlyForwards` and having a second ctor that accepts an `int ticks` to replace `FlyTimed`?

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs]: > FlyOffMap.cs(44,37): error SA1021: Negative sign should not be followed by a space.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs]: Whops, fixed.

**Comment by TBAALi-ros** [comment]
I don't TBAALi 邮箱：jiaxx903@tom.com 签名由 网易邮箱大师 定制 On 04/18/2020 00:09, abcdefg30 wrote: @abcdefg30 commented on this pull request. In OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs: > @@ -41,15 +41,15 @@ protected override void OnFirstRun(Actor self)
 			if (hasTarget)
 			{
 				QueueChild(new Fly(self, target));
-				QueueChild(new FlyTimed(-1, self));
+				QueueChild(new FlyForward(self, - 1));
 Whops, fixed. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or unsubscribe.

File: OpenRA.Mods.Common/Activities/Air/FlyForward.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyForward.cs:27]: We could avoid duplicating these two lines by the two ctors call a private ctor that assigns these. This is a minor point and would actually increase the number of lines, but is IMO a good practice to follow.

File: OpenRA.Mods.Common/Activities/Air/FlyForward.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyForward.cs]: Can we have a comment somewhere saying that it is explicitly valid for `flyTicks` to be -1 to mean "keep flying until cancelled"?

**Review by pchote — COMMENTED** [review]
LGTM otherwise. We'll also need to remember to remove the testcases.

**Comment by abcdefg30** [comment]
Updated. I can remove the testcases once we have a second +1. (Except if people prefer to construct their own testcases?)

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: I'm not sure this will do for moving targets. The aircraft won't achieve the proper distance if the target is moving in the same direction as the attacker. This might be especially relevant for air-to-air attacks, but is still a slight regression for the current first-party use-cases (mig and yak). 

This is not an issue for the explicit strafe case, because there it is targeting a fixed spot.

I also don't think the new behaviour is necessarily that much more sensible as default behaviour for units that can slide but can't hover (units that can hover do not make use of this code path anyway). It is still a weird oscillating back-and-forth that isn't optimal given it's flight characteristics and weapon range. The most sensible default behaviour in this case would be a circle strafe, maintaining a fixed distance and constantly facing towards the target.

So I think this is a good change for the Strafe attack type, when you explicitly want this particular flight pattern, but not for the default.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Hm good point. I hadn't considered moving targets, so I'll have to think about this again. Are you sure about the oscillating behaviour though? (IIrc I set the MiG testcase up for this branch, but don't remember.)

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: The behaviour for the MiG here is as follows:
- It first moves towards the target, facing it&mdash;so far so good.
- It then moves away from the target in a straight line, facing _away_ from the target so it cannot fire even though it is able to slide in any direction and should just be facing the target. 
- As soon as it reaches the exitRange it is immediately jerked back towards the target&mdash;still facing away from the target!&mdash;slowly turning and only facing the target when it is already almost on top of the target.

Steps 2 and 3 then repeat ad infinitum, meaning that in a sustained attack (or an attack started while already within weapon range) the plane is actually unable to fire it's weapon during most of the attack run while simultaneously being always closer than it needs to be thereby unnecessarily exposing itself.

So it is actually strictly worse than the behaviour shown in the gif, since in that case it is at least always facing the target and neither of the maneuvers are impossible given the defined flight characteristics.

As I said, the most logical behaviour would be to attempt a circle-strafe maneuver, but that requires some more thought. A simple improvement might be to do this instead:
```C#
QueueChild(new Fly(self, target, target.CenterPosition));
QueueChild(new FlyForward(self, 1));
QueueChild(new Fly(self, target, exitRange, WDist.MaxValue, target.CenterPosition));
```
This is similar to the behaviour I described above, except it will try to face the target while doing it. (It kinda looks like something you might see in Battlestar Galactica actually. :smile: )

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:279]: Thinking some more about it, `StrafeAttackRun` would need some adjustment as well for sliding aircraft. As I said above, the sliding aircraft will still be busy turning towards the target when it is already moving back towards the target. In order to account for that in the case when you explicitly want a strafe, we would need to add an extra distance on top of `exitRange` so that the aircraft is fully facing the target by the time it reaches the required range.

Something like
```C#
exitRange += aircraft.Info.Speed * 128 / aircraft.Info.TurnSpeed;
```
would do. We could of course require modders to just think of this themselves and set it manually, but this way it is actually more consistent with non-sliding aircraft which also do their turning outside of the `exitRange` and it also plays nice with the default value (the maximum weapon range).

I'm thinking that for aircraft that can slide _and_ can hover it would make more sense to just turn in place between strafing runs instead.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:279]: Actually, something like this would be even better:
```C#
QueueChild(new Fly(self, target, target.CenterPosition));
QueueChild(new FlyForward(self, exitRange));
QueueChild(new Fly(self, target, exitRange + new WDist(aircraft.Info.Speed * 64 / aircraft.Info.TurnSpeed), WDist.MaxValue, target.CenterPosition));
```

**Comment by pchote** [comment]
Where are we at with this? We will need to fix this before/if we want to implement the classic banshee/orca bomber behaviour in TS, which would have gone well with #18231.

**Comment by tovl** [comment]
Banshees and Orca bombers are not sliders (neither were they in the original), so this PR isn't relevant to them.

I already provided a basically copy-pastable solution to the problems I pointed out, so there is nothing else left for me to say here.

**Comment by tovl** [comment]
> Banshees and Orca bombers are not sliders (neither were they in the original), so this PR isn't relevant to them.

Closer inspection reveals this isn't quite true. They did slide, but only slightly.

**Comment by abcdefg30** [comment]
Sorry for leaving this dormant for so long. Updated with @tovl's suggestions which look/feel better ingame.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:230]: This won't necessarily be true, depending on the relative direction and speed of the two units. What will happen if this assumption fails?

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: Can we have a comment explaining the meaning of `aircraft.Info.Speed * 256 / aircraft.Info.TurnSpeed.Angle`?

File: OpenRA.Mods.Common/Activities/Air/FlyForward.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyForward.cs]: Style thought: Maybe define a default `ticks = -1` to make `FlyOffMap` a little cleaner?

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (abcdefg30) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:230]: IIrc: It'll fly "backwards" away from the target (still facing it if the aircraft can slide) which is the trade-off of using @tovl's suggestion here instead of always flying the distance straight. The other option I can think of would be giving `FlyStep` a target which it tries to face while flying the step.

**Comment by abcdefg30** [comment]
Updated. https://github.com/OpenRA/OpenRA/pull/17935#discussion_r470964761 is still open though.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:230]: If the assumption fails, it won't even reach this stage because it is still trying to catch up with the target?

Tbh, I don't think we should be breaking our heads over edge cases like this now, when #18386 and its predecessors will be changing the underlying assumptions anyway.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:230]: >  it won't even reach this stage because it is still trying to catch up with the target?

Again, this assumes that both units are moving in the same direction and do not turn. I raised the point specifically because "target changes direction while moving" is the kind of thing that is easily forgotten in constructed test cases but happens all the time during real gameplay.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:230]: This specific addition can't make things any worse than the current behaviour, so its not worth blocking this over it.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Activities/Air/FlyForward.cs
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: This will not give the correct distance when the aircraft gets repulsed along the way. This could be sidestepped by comparing the actual current position with the position on the previous tick instead.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: Depends on what we want to do, let the aircraft *fly forward* or *cover a distance*?

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: If the point is not to cover a set distance, then we shouldn't allow constructing with a distance argument in the first place. 

If the point is just to fly forward for some number of ticks&mdash;even when repulsion forces may in the extreme case make the aircraft barely move at all&mdash;then just define a number of ticks and don't bother.

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: Measuring the actual distance is an easy fix - rename `remainingDistance` to `distance` and make it readonly, add an `OnFirstRun` that sets a `WPos distanceFrom` variable, then check `(self.CenterPosition - distanceFrom).HorizontalLengthSquared >= distance.LengthSquared`

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: That would be measuring a different kind of distance: The cartesian distance from the starting point instead of the distance travelled along the trajectory. Tbh, I'm not sure which one would be better, but that wasn't the issue.

The issue is _not_ that the trajectory could be bent by repulsion&mdash;I'd assume from the code that abc intended to measure the bent trajectory; the issue is that because `FlyStep` doesn't take repulsion into account, this code doesn't _actually_ measure the distance travelled along the trajectory either!

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: The code that uses `FlyForwards` isn't smart enough to handle repulsion either, so i'm not sure it is worth holding up the bugfix over this. The current implementation seems fine, IMO.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: I'll try to explain myself again. Imo this activity is just about the plane's perspective ("I stupidly propel forwards"), not what happens around it. If repulsion pushed the plane half across the map, should we fly all the way back? (Distance delta would get negative.) This would then become a "fly to target" activity. (Or when just measuring trajectory, we'd need to ask ourselves if it still makes sense to fly forward; for example when being pushed forwards.) Imo not the goal here.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/FlyForward.cs:58]: > 
> 
> If the point is not to cover a set distance, then we shouldn't allow constructing with a distance argument in the first place.
> 
> If the point is just to fly forward for some number of ticks—even when repulsion forces may in the extreme case make the aircraft barely move at all—then just define a number of ticks and don't bother.

The distance-based constructor is just actively misleading then.

**Comment by pchote** [comment]
How do we want to proceed here?

**Comment by abcdefg30** [comment]
I don't think I have anything to change here right now.

**Comment by pchote** [comment]
I have already 👍'd so it would be good to get another review here.

**Comment by pchote** [comment]
This has now been over a month without any further review activity, so if you can resolve the rebase conflict and nobody else comments before this weekend with further discussion (or a request for more time to look) then i'll merge this.

**Comment by abcdefg30** [comment]
Rebased.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/255a36ff115656782e4d4df7c8a3128a7825d92b

**Pull request merged**
